File: LeetCode/[E] - 872. Leaf-Similar Trees/solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def leafSimilar(
        self, root1: Optional[TreeNode], root2: Optional[TreeNode]
    ) -> bool:
        leafes1, leafes2 = [], []
        self.leafes_collect(root1, leafes1)
        self.leafes_collect(root2, leafes2)
        return leafes1 == leafes2

    def leafes_collect(self, node, leafes):
        if not node:
            return
        if not node.left and not node.right:
            return leafes.append(node.val)
        self.leafes_collect(node.left, leafes)
        self.leafes_collect(node.right, leafes)

File: LeetCode/[E] - 872. Leaf-Similar Trees/solution.go
package main

import "slices"

// TreeNode - Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	leafes1, leafes2 := make([]int, 0, 100), make([]int, 0, 100)
	leafesCollect(root1, &leafes1)
	leafesCollect(root2, &leafes2)
	return slices.Equal(leafes1, leafes2)
}

func leafesCollect(node *TreeNode, leafes *[]int) {
	if node == nil {
		return
	}
	if node.Left == nil && node.Right == nil {
		*leafes = append(*leafes, node.Val)
		return
	}
	leafesCollect(node.Left, leafes)
	leafesCollect(node.Right, leafes)
}
